hugegraph-api/src/main/java/com/baidu/hugegraph/api/graph/BatchAPI.java

```diff
@@ -71,4 +82,99 @@ public BatchAPI() {
+        public Map<String, Object> properties;
+        @JsonProperty("type")
+        public String type;
+
+        @Override
+        public abstract void checkCreate(boolean isBatch);
+
+        @Override
+        public abstract void checkUpdate();
+
+        protected abstract Object[] properties();
+    }
+
+    /**
+     * Get vertex/edge ID from JsonVerex/JsonEdge
+     * TODO: Should support AUTOMATIC ElementID ?
+     * TODO: How to get JsonEdge's fields ?
+     */
+    protected String getElementId(HugeGraph g, String labelId,
+                                  JsonElement element, boolean isVertex) {
+        if (isVertex) {
+            List<Id> pkIds = g.vertexLabel(element.label).primaryKeys();
+            List<Object> pkValues = new ArrayList<>(pkIds.size());
+            pkIds.forEach(pkId -> {
+                String propertyKey = g.propertyKey(pkId).name();
+                Object propertyValue = element.properties.get(propertyKey);
+                E.checkState(propertyValue != null,
+                             "The value of primary key '%s' can't be null",
+                             g.propertyKey(pkId).name());
+                pkValues.add(propertyValue);
+            });
+
+            String name = SplicingIdGenerator.concatValues(pkValues);
+            return SplicingIdGenerator.splicing(labelId, name).asString();
+        } else {
+            // TODO: Merge getEdgeId() from EdgeAPI to here
+            // EdgeId edgeId = new EdgeId(); // need jsonEdge here
+        }
+        return null;
+    }
+
+    protected void updateExistElement(JsonElement oldElement,
+                                      JsonElement newElement,
+                                      Map<String, UpdateStrategy> strategy) {
+        if (oldElement != null && newElement != null) {
+            strategy.entrySet().forEach(kv -> {
+                String key = kv.getKey();
+                if (oldElement.properties.get(key) != null) {
+                    Object value = UpdateStrategy.checkAndUpdateProperty(
+                            oldElement.properties.get(key),
+                            newElement.properties.get(key),
+                            kv.getValue());
+                    newElement.properties.put(key, value);
+                }
+            });
+        }
+    }
+
+    // TODO: Combine multi update methond into one (Element & JsonElement)
+    protected void updateExistElement(Element oldElement, JsonElement element,
+                                      Map<String, UpdateStrategy> strategy) {
+        if (oldElement != null && element != null) {
+            strategy.entrySet().forEach(kv -> {
+                String key = kv.getKey();
+                if (oldElement.property(key).isPresent()) {
+                    Object value = UpdateStrategy.checkAndUpdateProperty(
+                            oldElement.property(key).value(),
+                            element.properties.get(key),
```

Comment:
let new property value be converted by property-key like this [pkey.convValue(value, false)](https://github.com/hugegraph/hugegraph/blob/master/hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/TraversalUtil.java#L569)

how to get property-key: `graph.propertyKey(name)`

Reply:
> 
> 
> let new property value be converted by property-key like this [pkey.convValue(value, false)](https://github.com/hugegraph/hugegraph/blob/master/hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/TraversalUtil.java#L569)
> 
> how to get property-key: `graph.propertyKey(name)`

is it like `value = g.propertyKey(key).convValue(value, false)` ? Wonder why need to convert pkValue again? or just for safety?

PS: disappear for some days, and back now~   (will fix them soon)